### PR TITLE
Fix S1144 and S1172: handle new main syntax

### LINKED ISSUE #888: Fix S1144: False positive on async Main
A false positive S1144 is thrown on the line: 
```csharp 
static async Task Main(string[] args)
```

In C# 7.1 this is a valid entrypoint method. https://blogs.msdn.microsoft.com/mazhou/2017/05/30/c-7-series-part-2-async-main/

```csharp
using System;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Test");
        }
    }
}
```

### LINKED ISSUE #1024: Fix S1172: Rule should also ignore the new Main syntax
### Description

The rule doesn't report when the code doesn't use `args` from Main method.

### Repro steps

```
using System;
using System.Threading.Tasks;

namespace ConsoleApp1
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Console.WriteLine("Test");
        }
    }
}
```

### Expected behavior

The rule doesn't raise an issue when not using `args`.

### Actual behavior

The rule raises an issue.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #888 and fix #1024

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs [hunk 1]
```diff
@@ -30,12 +30,19 @@ public static class KnownMethods
...
-                (methodSymbol.Parameters.Length == 0 ||
-                    (methodSymbol.Parameters.Length == 1 && methodSymbol.Parameters[0].IsType(KnownType.System_String_Array)));
+                (
+                    methodSymbol.Parameters.Length == 0 ||
+                    (methodSymbol.Parameters.Length == 1 && methodSymbol.Parameters[0].IsType(KnownType.System_String_Array))
+                ) &&
+                (
+                    methodSymbol.ReturnsVoid ||
+                    methodSymbol.ReturnType.Is(KnownType.System_Int32) ||
+                    (methodSymbol.ReturnType.Is(KnownType.System_Threading_Tasks_Task) && methodSymbol.IsAsync)
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs]: Experimenting shows that the _async_ keyword is not required (although you would than have to explicitly return a _Task_ for the code to compile).

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/MethodParameterUnused.Fixed.cs [hunk 2]
```diff
@@ -53,6 +53,11 @@ public void Caller()
         {
         }
 
+        static async Task Main(string[] args) // Compliant - new main syntax
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/MethodParameterUnused.Fixed.cs]: Additional test cases:
 _Task<int>_ should be compliant
 _Task<string> should not

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs (hunk 1 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/KnownMethods.cs]: Good catch! I will remove the `async` check as it is not stated by the blog post.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/MethodParameterUnused.Fixed.cs (hunk 2 above)
Comment (Evangelink) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/MethodParameterUnused.Fixed.cs]: Good catch too. I will add `Task<int>`.

**Pull request merged**